src/protocols/bdx/tests/TestBdxTransferSession.cpp

```diff
@@ -113,183 +154,490 @@ void VerifyStatusReport(nlTestSuite * inSuite, void * inContext, const System::P
-            }
-        }
-        else
-        {
-            NL_TEST_ASSERT(inSuite, false); // Metadata length mismatch
-        }
+        Reset();
     }
-
-    // Verify that MaxBlockSize was set appropriately
-    NL_TEST_ASSERT(inSuite, acceptReceiver.GetTransferBlockSize() <= initData.MaxBlockSize);
 }
 
-// Helper method for preparing a sending a BlockQuery message between two TransferSession objects.
-void SendAndVerifyQuery(nlTestSuite * inSuite, void * inContext, TransferSession & queryReceiver, TransferSession & querySender,
-                        TransferSession::OutputEvent & outEvent)
+// Helper method for preparing and sending a BlockQuery message between two TransferSession objects.
+void SendAndVerifyQuery(nlTestSuite * inSuite, TransferSession & queryReceiver, TransferSession & querySender)
 {
+    // Set the expected output event message
+    sInitiatorExpectedOutputEvent.msgTypeData.MessageType = static_cast<uint8_t>(MessageType::BlockQuery);
+
+    sResponderExpectedOutputEvent.EventType = TransferSession::OutputEventType::kQueryReceived;
     // Verify that querySender emits BlockQuery message
     CHIP_ERROR err = querySender.PrepareBlockQuery();
     NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
-    querySender.PollOutput(outEvent, kNoAdvanceTime);
-    NL_TEST_ASSERT(inSuite, outEvent.EventType == TransferSession::OutputEventType::kMsgToSend);
-    VerifyBdxMessageToSend(inSuite, inContext, outEvent, MessageType::BlockQuery);
-    VerifyNoMoreOutput(inSuite, inContext, querySender);
+}
 
-    // Pass BlockQuery to queryReceiver and verify queryReceiver emits QueryReceived event
-    err = AttachHeaderAndSend(outEvent.msgTypeData, std::move(outEvent.MsgData), queryReceiver);
+// Helper method for preparing and sending the same block data to simulate a duplicate block error and also setting up the expected
+// output event for the responder to kBlockReceived for the first block counter and kStatusReceived for the duplicate block
+void SendAndVerifyStatusReportForSameBlock(nlTestSuite * inSuite, TransferSession & sender, TransferSession & receiver, bool isEof,
+                                           uint32_t inBlockCounter, TransferSession::OutputEvent & event)
+{
+    CHIP_ERROR err       = CHIP_NO_ERROR;
+    uint8_t fakeData[64] = { 0 };
+    uint8_t fakeDataLen  = sizeof(fakeData);
+
+    TransferSession::BlockData blockData;
+    blockData.Data   = fakeData;
+    blockData.Length = fakeDataLen;
+    blockData.IsEof  = false;
+
+    sInitiatorExpectedOutputEvent.msgTypeData.MessageType =
+        static_cast<uint8_t>(isEof ? MessageType::BlockEOF : MessageType::Block);
+    // Set the expected BDX event type
+    if (inBlockCounter == 0)
+    {
+        sResponderExpectedOutputEvent.EventType = TransferSession::OutputEventType::kBlockReceived;
+        sResponderExpectedOutputEvent.blockdata = blockData;
```

Comment:
The blockData will be used after scope (since it is **non-owning** and pointing to a stack buffer). Please fix.